File: js/NativeGodotModule.ts
/**************************************************************************/
/** biome-ignore-all lint/suspicious/noRedeclare: native globals are declared for the JSI install state. */
/** biome-ignore-all lint/suspicious/noConsole: native module install diagnostics are logged to the JS console. */
/*  NativeGodotModule.ts                                                  */
/**************************************************************************/
/* Copyright (c) 2024-2025 Slay GmbH                                      */
/*                                                                        */
/* Permission is hereby granted, free of charge, to any person obtaining  */
/* a copy of this software and associated documentation files (the        */
/* "Software"), to deal in the Software without restriction, including    */
/* without limitation the rights to use, copy, modify, merge, publish,    */
/* distribute, sublicense, and/or sell copies of the Software, and to     */
/* permit persons to whom the Software is furnished to do so, subject to  */
/* the following conditions:                                              */
/*                                                                        */
/* The above copyright notice and this permission notice shall be         */
/* included in all copies or substantial portions of the Software.        */
/*                                                                        */
/* THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,        */
/* EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF     */
/* MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. */
/* IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY   */
/* CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,   */
/* TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE      */
/* SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.                 */
/**************************************************************************/

import {type TurboModule, TurboModuleRegistry} from "react-native";
import {createWorkletRuntime, runOnRuntimeAsync, type WorkletRuntime} from "react-native-worklets";

export interface Spec extends TurboModule {
  installTurboModule(): boolean;
}

const GodotInstaller = TurboModuleRegistry.getEnforcing<Spec>("NativeGodotModule");

export interface GodotModuleInterface {
  createInstance(args: Array<string>): any;
  getInstance(): any;
  API(): any;
  updateWindow(windowName: string): any;
  pause(): void;
  resume(): void;
  is_paused(): boolean;
  runOnGodotThread<T>(f: () => T): Promise<T>;
  createGodotQueue(): object;
  destroyInstance(): void;
  crash(): void;
}

console.log("Loading NativeGodotModule...");

declare global {
  var RTNGodot: GodotModuleInterface | undefined; // Godot
  var __godotWorkletRuntime: WorkletRuntime | undefined;
}

const getInstalledGodotModule = () => globalThis.RTNGodot;

if (globalThis.RTNGodot == null) {
  if (GodotInstaller == null || typeof GodotInstaller.installTurboModule !== "function") {
    console.error(
      "Native Godot Module cannot be found! Make sure you correctly " +
        "installed native dependencies and rebuilt your app.",
    );
  } else {
    console.log("Calling NativeGodotModule.installTurboModule()");
    const result = GodotInstaller.installTurboModule();
    if (!result) {
      console.log("Failed NativeGodotModule.installTurboModule()");
    } else {
      const RTNGodot = getInstalledGodotModule();
      if (RTNGodot == null) {
        console.log("Failed NativeGodotModule.installTurboModule(): RTNGodot was not installed");
      } else {
        const godotQueue = RTNGodot.createGodotQueue();
        globalThis.__godotWorkletRuntime = createWorkletRuntime({
          name: "ReactNativeGodot",
          queue: godotQueue,
          initializer: () => {
            "worklet";
            globalThis.RTNGodot = RTNGodot;
          },
        });
      }
    }
  }

  const RTNGodot = getInstalledGodotModule();
  if (RTNGodot == null) {
    console.log(`Unable to load NativeGodotModule: ${RTNGodot}`);
  }
} else {
  console.log("NativeGodotModule loaded.");
}

export const RTNGodot = globalThis.RTNGodot as GodotModuleInterface;

export function runOnGodotThread<T>(f: () => T): Promise<T> {
  console.log("Calling: runOnGodotThread");
  const runtime = globalThis.__godotWorkletRuntime;
  if (runtime == null) {
    return Promise.reject(new Error("NativeGodotModule worklet runtime is not installed"));
  }
  return runOnRuntimeAsync(runtime, f);
}

File: android/src/main/java/net.somesoap/rtngodot/NativeGodotModule.java
/**************************************************************************/
/*  NativeGodotModule.java                                                */
/**************************************************************************/
/* Copyright (c) 2024-2025 Slay GmbH                                      */
/*                                                                        */
/* Permission is hereby granted, free of charge, to any person obtaining  */
/* a copy of this software and associated documentation files (the        */
/* "Software"), to deal in the Software without restriction, including    */
/* without limitation the rights to use, copy, modify, merge, publish,    */
/* distribute, sublicense, and/or sell copies of the Software, and to     */
/* permit persons to whom the Software is furnished to do so, subject to  */
/* the following conditions:                                              */
/*                                                                        */
/* The above copyright notice and this permission notice shall be         */
/* included in all copies or substantial portions of the Software.        */
/*                                                                        */
/* THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,        */
/* EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF     */
/* MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. */
/* IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY   */
/* CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,   */
/* TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE      */
/* SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.                 */
/**************************************************************************/

package net.somesoap.rtngodot;

import androidx.annotation.OptIn;

import java.util.Objects;

import com.facebook.jni.HybridData;
import com.facebook.proguard.annotations.DoNotStrip;
import com.facebook.react.bridge.ReactApplicationContext;
import com.facebook.react.bridge.ReactMethod;
import com.facebook.react.common.annotations.FrameworkAPI;
import com.facebook.react.module.annotations.ReactModule;
import com.facebook.react.turbomodule.core.CallInvokerHolderImpl;
import com.migeran.NativeGodotModuleSpec;

@OptIn(markerClass = FrameworkAPI.class)
@ReactModule(name = NativeGodotModule.NAME)
public class NativeGodotModule extends NativeGodotModuleSpec {
	public static final String NAME = "NativeGodotModule";

	@DoNotStrip
	@SuppressWarnings("unused")
	private final HybridData mHybridData;

	public NativeGodotModule(ReactApplicationContext context) {
		super(context);
		CallInvokerHolderImpl holder =
				(CallInvokerHolderImpl)context.getCatalystInstance().getJSCallInvokerHolder();
		mHybridData = initHybrid(
				Objects.requireNonNull(context.getJavaScriptContextHolder()).get(),
				holder);
	}

	private native HybridData initHybrid(long jsContext, CallInvokerHolderImpl jsCallInvokerHolder);

	@ReactMethod(isBlockingSynchronousMethod = true)
	@Override
	public native boolean installTurboModule();
}
